src/ralph/backends/data/s3.py

```diff
@@ -0,0 +1,360 @@
+            target (str or None): The target bucket and the target object
+                separated by a `/`.
+                If target is `None`, the default bucket is used and a random
+                (uuid4) object is created.
+                If target does not contain a `/`, it is assumed to be the
+                target object and the default bucket is used.
+            chunk_size (int or None): Ignored.
+            ignore_errors (bool): Ignored.
+            operation_type (BaseOperationType or None): The mode of the write
+                operation.
+                If operation_type is `CREATE` or `INDEX`, the target object is
+                expected to be absent. If the target object exists a
+                `BackendException` is raised.
+
+        Return:
+            int: The number of written objects.
+
+        Raise:
+            BackendException: If a failure during the write operation occurs.
+            BackendParameterException: If a backend argument value is not valid.
+        """
+        try:
+            first_record = next(iter(data))
+        except StopIteration:
+            logger.info("Data Iterator is empty; skipping write to target.")
+            return 0
+
+        if not operation_type:
+            operation_type = self.default_operation_type
+
+        if not target:
+            target = f"{self.default_bucket_name}/{now()}-{uuid4()}"
+            logger.info(
+                "Target not specified; using default bucket with random file name: %s",
+                target,
+            )
+
+        elif "/" not in target:
+            target = f"{self.default_bucket_name}/{target}"
+            logger.info(
+                "Target not specified; using default bucket: %s",
+                target,
+            )
+
+        target_bucket, target_object = target.split("/", 1)
+
+        if operation_type in [
+            BaseOperationType.APPEND,
+            BaseOperationType.DELETE,
+            BaseOperationType.UPDATE,
+        ]:
+            msg = "%s operation_type is not allowed."
+            logger.error(msg, operation_type.name)
+            raise BackendParameterException(msg % operation_type.name)
+
+        if target_object in list(self.list(target=target_bucket)):
+            msg = "%s already exists and overwrite is not allowed for operation %s"
+            logger.error(msg, target_object, operation_type)
+            raise BackendException(msg % (target_object, operation_type))
+
+        logger.debug("Creating archive: %s", target_object)
+
+        count = 0
+        file_obj = BytesIO()
+        if isinstance(first_record, dict):
+            for chunk in data:
+                file_obj.write(
+                    bytes(f"{json.dumps(chunk)}\n", encoding=self.locale_encoding)
+                )
+                count += 1
+        else:
+            for chunk in data:
+                file_obj.write(chunk)
+                count += 1
+
+        file_obj.seek(0)
```

Comment:
Storing a copy of the whole file in memory (`BytesIO`) might be problematic when working with large files.
I think it would be preferable to write the `data` content to a temporary file on disk and then use it with `upload_file`.
Alternatively, maybe we could implement a class with a `read` method wrapping the `data` iterable?
WDYT?

Also, `data` could be an instance of `IOBase`, if that's the case (and it contains binary data), could we pass it directly to `upload_fileobj`?

Reply:
Using now `StreamingIterator` from `requests-toolbelt`, to be able to use `upload_fileobj`, let me know what you think of it!